{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LAB9"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write a local-search algorithm (eg. an EA) able to solve the Problem instances 1, 2, 5, and 10 on a 1000-loci genomes, using a minimum number of fitness calls. That's all.\n",
    "\n",
    "#### Deadlines:\n",
    "- Submission: Sunday, December 3 (CET)\n",
    "- Reviews: Sunday, December 10 (CET)\n",
    "\n",
    "Notes:\n",
    "\n",
    "- Reviews will be assigned on Monday, December 4\n",
    "- You need to commit in order to be selected as a reviewer (ie. better to commit an empty work than not to commit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from random import choices, choice, random, randint, sample\n",
    "from dataclasses import dataclass\n",
    "from copy import copy\n",
    "from tqdm import tqdm\n",
    "import numpy as np\n",
    "from pprint import pprint\n",
    "import matplotlib.pyplot as plt\n",
    "import lab9_lib"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genetic Algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENOME_SIZE = 1000\n",
    "POPULATION_SIZE = 600\n",
    "OFFSPRING_SIZE = 450\n",
    "TOURNAMENT_SIZE = 2\n",
    "MUTATION_PROBABILITY = .3\n",
    "CROSSOVER_RATE = .8\n",
    "NUM_GENERATIONS = 3000\n",
    "PROBLEM_SIZE = 10\n",
    "tournament = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "fitness = lab9_lib.make_problem(PROBLEM_SIZE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "00000101010100111001101100011101110101001001011011: 50.00%\n",
      "11001111101111101111011101001101101110000010110110: 64.00%\n",
      "10110100100011111001011000100101011101100100001001: 48.00%\n",
      "00010000000000101001010100001101111111101101110010: 44.00%\n",
      "10011000011110011000011110111110111111111000010010: 58.00%\n",
      "00111011100011000011011010000010101101101010001100: 46.00%\n",
      "11001010011010010011100000111000101101001010011111: 50.00%\n",
      "00111000010001111100101101110110010100011001010110: 50.00%\n",
      "01101110111100111111111111101110010010100000010000: 58.00%\n",
      "10001100010010001100100111100010100110101100111100: 46.00%\n",
      "10\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for n in range(10):\n",
    "    ind = choices([0, 1], k=50)\n",
    "    print(f\"{''.join(str(g) for g in ind)}: {fitness(ind):.2%}\")\n",
    "\n",
    "print(fitness.calls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "@dataclass\n",
    "class Individual:\n",
    "    genotype: list[int] \n",
    "    fitness: tuple \n",
    "\n",
    "def select_parent(population):\n",
    "\n",
    "    if tournament :\n",
    "        \n",
    "        \n",
    "        pool = [choice(population) for _ in range(TOURNAMENT_SIZE)]\n",
    "        champion = max(pool, key=lambda i: i.fitness)\n",
    "            \n",
    "        \n",
    "        \n",
    "\n",
    "    else:\n",
    "        #roulette wheel\n",
    "        fitness_scores = [fitness(ind.genotype) for ind in population]\n",
    "        total_fitness = sum(fitness_scores)\n",
    "        probabilities = [score / total_fitness for score in fitness_scores]\n",
    "        selected_index = choices(range(len(population)), weights=probabilities)[0]\n",
    "        champion = population[selected_index]\n",
    "    \n",
    "    return champion\n",
    "\n",
    "# mutation strategies\n",
    "\n",
    "#select a locus in the genotype and flip the correpsonding bit\n",
    "def mutate(ind: Individual) -> Individual:\n",
    "    \n",
    "    offspring= copy(ind)\n",
    "    pos = randint(0, GENOME_SIZE-1)\n",
    "    \n",
    "    offspring.genotype[pos] = 1 - offspring.genotype[pos]\n",
    "    offspring.fitness = None\n",
    "    return  offspring\n",
    "\n",
    "#for each locus in the genotype flip the correpsonding bit with probability MUTATION_PROBABILITY\n",
    "def mutate_all(ind: Individual)-> Individual:\n",
    "    mutated_individual = copy(ind)\n",
    "    for i in range(len(mutated_individual.genotype)):\n",
    "        if random() < MUTATION_PROBABILITY:\n",
    "            mutated_individual.genotype[i] = 1 - mutated_individual.genotype[i]  # Flip bit\n",
    "    return mutated_individual\n",
    "\n",
    "#shift each allele right of one position (starting from the last position) and assign to the first postition \n",
    "# 0 or 1 with equal probability\n",
    "def mutate_shift(ind: Individual)-> Individual:\n",
    "    mutated_individual = copy(ind)\n",
    "    i=len(mutated_individual.genotype)-1\n",
    "    while i >= 0:\n",
    "        if i == 0:\n",
    "            if random() < 0.5:\n",
    "                mutated_individual.genotype[i] = 1\n",
    "            else:\n",
    "                mutated_individual.genotype[i] = 0\n",
    "            break\n",
    "        mutated_individual.genotype[i] = mutated_individual.genotype[i-1]  # shift right by one bit\n",
    "        \n",
    "        i-=1\n",
    "\n",
    "    return mutated_individual\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Crossover startegies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def one_cut_xover(ind1: Individual, ind2: Individual) -> Individual:\n",
    "    cut_point = randint(0, GENOME_SIZE-1)\n",
    "    offspring1 = Individual(fitness=None,genotype=ind1.genotype[:cut_point]+ind2.genotype[cut_point:])\n",
    "    offspring2 = Individual(fitness=None,genotype=ind2.genotype[:cut_point]+ind1.genotype[cut_point:])\n",
    "    \n",
    "    return offspring1 , offspring2\n",
    "\n",
    "def  two_cut_xover(ind1: Individual, ind2: Individual)-> Individual:\n",
    "    point1 = randint(0, GENOME_SIZE-1)\n",
    "    point2 = randint(0, GENOME_SIZE-1)\n",
    "    if point2 > point1:\n",
    "        offspring1 = Individual(fitness=None,genotype=ind1.genotype[:point1] + ind2.genotype[point1:point2] + ind1.genotype[point2:])\n",
    "        offspring2 = Individual(fitness=None,genotype=ind2.genotype[:point1] + ind1.genotype[point1:point2] + ind2.genotype[point2:])\n",
    "    else:\n",
    "        offspring1 = Individual(fitness=None,genotype=ind1.genotype[:point2] + ind2.genotype[point2:point1] + ind1.genotype[point1:])\n",
    "        offspring2 = Individual(fitness=None,genotype=ind2.genotype[:point2] + ind1.genotype[point2:point1] + ind2.genotype[point1:])\n",
    "    return offspring1, offspring2\n",
    "\n",
    "\n",
    "def uniform_crossover(ind1: Individual, ind2: Individual)-> Individual:\n",
    "    child1 = Individual(fitness=None,genotype=[])\n",
    "    child2 = Individual(fitness=None,genotype=[])\n",
    "    \n",
    "    for gene1, gene2 in zip(ind1.genotype, ind2.genotype):\n",
    "        if random() < 0.5:\n",
    "            child1.genotype.append(gene1)\n",
    "            child2.genotype.append(gene2)\n",
    "        else:\n",
    "            child1.genotype.append(gene2)\n",
    "            child2.genotype.append(gene1)\n",
    "    return child1, child2\n",
    "\n",
    "def n_cut_crossover(parent1: Individual, parent2: Individual)-> Individual:\n",
    "    # Sorting the cut points in ascending order\n",
    "    n=(GENOME_SIZE//PROBLEM_SIZE) -1\n",
    "    cut_points = sorted(sample(range(1, len(parent1.genotype)), n))\n",
    "\n",
    "    child1 = Individual(fitness=None,genotype=[])\n",
    "    child2 = Individual(fitness=None,genotype=[])\n",
    "\n",
    "    parents = [parent1,parent2]\n",
    "    childs = [child1,child2]\n",
    "    current_parent= 0\n",
    "    for i in range(len(childs)):\n",
    "        childs[i].genotype.extend(parents[current_parent].genotype[:cut_points[0]])\n",
    "        current_parent= 1- current_parent\n",
    "        for j in range(len(cut_points)):\n",
    "            if j == (len(cut_points)-1):\n",
    "                childs[i].genotype.extend(parents[current_parent].genotype[cut_points[j]:])\n",
    "                if len(cut_points) % 2 == 0:\n",
    "                    current_parent= 1- current_parent\n",
    "                    continue\n",
    "                continue\n",
    "            childs[i].genotype.extend(parents[current_parent].genotype[cut_points[j]:cut_points[j+1]])\n",
    "            current_parent= 1- current_parent\n",
    "\n",
    "    return child1, child2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def initialize_population():\n",
    "    \n",
    "\n",
    "    population = [Individual(genotype=[choice((0,1)) for _ in range(GENOME_SIZE)],fitness=None) for _ in range(POPULATION_SIZE)]\n",
    "\n",
    "    for i in population:\n",
    "\n",
    "        i.fitness=fitness(i.genotype)\n",
    "\n",
    "    print(max(population, key=lambda x : x.fitness).fitness)\n",
    "\n",
    "    return population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "best = [[],0.0]\n",
    "num_iterations=0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ga(generational, xover_and_mutation, early_stopping):\n",
    "\n",
    "    global POPULATION_SIZE\n",
    "    global OFFSPRING_SIZE \n",
    "    global MUTATION_PROBABILITY \n",
    "    global CROSSOVER_RATE \n",
    "    global NUM_GENERATIONS \n",
    "    global PROBLEM_SIZE \n",
    "    global TOURNAMENT_SIZE\n",
    "    global niching\n",
    "    \n",
    "    best_fitnesses=[]\n",
    "    num_iterations=0\n",
    "    population= initialize_population()\n",
    "    last_fit = 0.0\n",
    "    last_5h = 0.0\n",
    "    for generation in tqdm(range(NUM_GENERATIONS)):\n",
    "        offspring=list()\n",
    "        num_iterations+=1\n",
    "        for counter in range(OFFSPRING_SIZE//2):\n",
    "\n",
    "            new_off=list()\n",
    "            if xover_and_mutation:\n",
    "                #Xover AND mutation\n",
    "                p1=select_parent(population)\n",
    "                p2=select_parent(population)\n",
    "                parents=[p1,p2]\n",
    "                if random() < CROSSOVER_RATE:     #self adapt mutation probability\n",
    "                        #add more xovers \n",
    "                    o1,o2 = two_cut_xover(p1,p2) \n",
    "                    new_off.append(o1)\n",
    "                    new_off.append(o2)\n",
    "                    \n",
    "\n",
    "                if len(new_off):\n",
    "                    for i in range(len(new_off)):\n",
    "                        if random() < MUTATION_PROBABILITY:\n",
    "                            new_off[i]=mutate_shift(new_off[i])\n",
    "                \n",
    "                else :\n",
    "                    for i in range(len(parents)):\n",
    "                        new_off.append(mutate_shift(parents[i]))\n",
    "\n",
    "                for os in new_off:   \n",
    "                    offspring.append(os)\n",
    "            else:\n",
    "                #xover OR mutation\n",
    "                p1=select_parent(population)\n",
    "                p2=select_parent(population)\n",
    "                if random() < CROSSOVER_RATE:     \n",
    "                    \n",
    "                    o1,o2 = two_cut_xover(p1,p2)                                                     \n",
    "                    \n",
    "                else:\n",
    "                    \n",
    "                    o1 = mutate(p1)\n",
    "                    o2 = mutate(p2)\n",
    "                    \n",
    "                    \n",
    "                offspring.append(o1)\n",
    "                offspring.append(o2)\n",
    "        \n",
    "        for i in offspring:\n",
    "            i.fitness = fitness(i.genotype)\n",
    "        \n",
    "\n",
    "        if not generational :\n",
    "            population.extend(offspring)\n",
    "            population.sort(key= lambda i: i.fitness , reverse = True)\n",
    "            population = population[:POPULATION_SIZE]\n",
    "        \n",
    "        else:\n",
    "            #if generational the offspring completely replace the individuals of the past generation\n",
    "            offspring.sort(key= lambda i: i.fitness , reverse = True)\n",
    "            population = offspring[:POPULATION_SIZE]\n",
    "        \n",
    "        #keep track of the best fitness at every iteration to plot in the end\n",
    "        best_fitnesses.append(population[0].fitness)\n",
    "\n",
    "        #update the best absolute fitness found (best[0] is the genotype of the solution, \n",
    "        # best[1] is its fitness value)\n",
    "        if best :\n",
    "            if population[0].fitness > best[1]:\n",
    "                best[0] = population[0].genotype\n",
    "                best[1] = population[0].fitness\n",
    "        else:\n",
    "            best[0] = population[0].genotype\n",
    "            best[1] = population[0].fitness\n",
    "\n",
    "\n",
    "        #adaptive mutation rate: every 100 generations if there has been little to no upgrade in the \n",
    "        # best fitness value, increase the mutation probability\n",
    "        if (generation +1) % 100 == 0:\n",
    "            fit_diff = population[0].fitness - last_fit\n",
    "            last_fit = population[0].fitness\n",
    "            if fit_diff < 0.01:\n",
    "                MUTATION_PROBABILITY*=1.1\n",
    "        \n",
    "        #stop the algorithm if the best value of the fitness is reached (1) or if there has not been \n",
    "        # any (relevant) upgrade in the fitness value last 500 generations\n",
    "        if early_stopping:\n",
    "            if best[1]== 1.0 :\n",
    "                break\n",
    "            if (num_iterations +1) % 500 == 0:\n",
    "               fit_diff_5h = population[0].fitness - last_5h\n",
    "               last_5h = population[0].fitness\n",
    "               if fit_diff_5h < 0.01:\n",
    "                   break\n",
    "                    \n",
    "    #plot the best fitness value evolution through the generations\n",
    "    plt.plot(range(1, num_iterations + 1), best_fitnesses)\n",
    "    plt.title('Problem size : {}, Fitness calls: {}'.format(PROBLEM_SIZE,fitness.calls))\n",
    "    plt.xlabel('Generation')\n",
    "    plt.ylabel('Best Fitness')\n",
    "    plt.show()\n",
    "    print(fitness.calls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ga(generational=False, xover_and_mutation=True, early_stopping=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.406899 [1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1]\n"
     ]
    }
   ],
   "source": [
    "print(best[1],best[0])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
